### PR TITLE
Dynamic binding of bolt connector to available ports

### PR DESCRIPTION
**Pull request by MishaDemianenko**
Allow bolt connectors to perform dynamic binding to any free port
Add possibility to find out out which port was used to register any connector
to allow any further operations.

### CODE REVIEW

**Comment by MishaDemianenko** [comment]
@lutovich wdyt?

File: community/bolt/src/main/java/org/neo4j/bolt/transport/NettyServer.java [hunk 1]
```diff
@@ -46,8 +50,9 @@
     private static final int NUM_SELECTOR_THREADS = Math.max( 1, Integer.getInteger(
             "org.neo4j.selectorThreads", Runtime.getRuntime().availableProcessors() * 2 ) );
 
-    private final Collection<ProtocolInitializer> bootstrappers;
+    private final Map<BoltConnector, ProtocolInitializer> bootstrappersMap;
     private final ThreadFactory tf;
+    private BoltConnectorRegister connectionRegister;
```
Comment (lutovich) [inline community/bolt/src/main/java/org/neo4j/bolt/transport/NettyServer.java]: Maybe make field `final`?

File: community/kernel/src/main/java/org/neo4j/kernel/api/bolt/BoltConnectorRegister.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import org.neo4j.helpers.HostnamePort;
+
+/**
+ * Connector tracker that keeps information about local address that any configured connector get during bootstrapping.
+ */
+public class BoltConnectorRegister
+{
+
+    private ConcurrentHashMap<String,HostnamePort> connectorsInfo = new ConcurrentHashMap<>();
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/api/bolt/BoltConnectorRegister.java]: Maybe make field `final`?

File: community/kernel/src/main/java/org/neo4j/kernel/api/bolt/BoltConnectorRegister.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.neo4j.helpers.HostnamePort;
+
+/**
+ * Connector tracker that keeps information about local address that any configured connector get during bootstrapping.
+ */
+public class BoltConnectorRegister
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/api/bolt/BoltConnectorRegister.java]: What do you think about `BoltPortRegister` name?

**Review by lutovich — APPROVED** [review]
@MishaDemianenko changes look good! couple tiny comments.

**Comment by MishaDemianenko** [comment]
@lutovich make all those things final and rename the class. Not sure if that will make sense in future since most probably the same component will track HTTP ports for corresponding connectors, but it's  good for now :)

**Pull request merged**